From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nagarjuna Kristam <nkristam@nvidia.com>
Date: Tue, 4 Jul 2023 16:34:28 +0530
Subject: [PATCH 223/245] NVIDIA: SAUCE: PCI: tegra194: Remove debugfs if link
 up

Issue:
debugfs entry is available only when link state is up.
During shutdown debugfs remove API is called before link state check.
This results in duplicate API invocation during shutdown
sequence resulting in kernel panic.

Fix:
Move debugfs removal API post link state check.

Bug 4169798

Change-Id: Icbfa7288e25270e238139f183e5f336c612dfc58
Signed-off-by: Nagarjuna Kristam <nkristam@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-jammy/+/2930744
(cherry picked from commit b422df5164c44d17ece111a5e1bcead16fff9a6a)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-jammy/+/2932553
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2966524
Reviewed-by: Prathamesh Shete <pshete@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
Tested-by: Prathamesh Shete <pshete@nvidia.com>

diff --git a/drivers/pci/controller/dwc/pcie-tegra194.c b/drivers/pci/controller/dwc/pcie-tegra194.c
index 02fd890b3ac02..9792b055cd18c 100644
--- a/drivers/pci/controller/dwc/pcie-tegra194.c
+++ b/drivers/pci/controller/dwc/pcie-tegra194.c
@@ -2623,7 +2623,6 @@ static void tegra_pcie_dw_shutdown(struct platform_device *pdev)
 	struct tegra_pcie_dw *pcie = platform_get_drvdata(pdev);
 
 	if (pcie->of_data->mode == DW_PCIE_RC_TYPE) {
-		debugfs_remove_recursive(pcie->debugfs);
 		disable_irq(pcie->prsnt_irq);
 		disable_irq(pcie->pci.pp.irq);
 		if (IS_ENABLED(CONFIG_PCI_MSI))
@@ -2632,6 +2631,8 @@ static void tegra_pcie_dw_shutdown(struct platform_device *pdev)
 			return;
 		if (!pm_runtime_enabled(pcie->dev))
 			return;
+
+		debugfs_remove_recursive(pcie->debugfs);
 		tegra_pcie_dw_pme_turnoff(pcie);
 		tegra_pcie_unconfig_controller(pcie);
 		pm_runtime_put_sync(pcie->dev);
-- 
2.43.0

